Adds first draft of Confirm a Deposit to Exchange (#3)
Annamarie2019/deposit: confirm-deposit.md

# Confirm a Deposit to the Exchange

Deposits are made in several stages. 
One of the stages is to go through the [Exchange server](https://docs.flare.network/exchange/architecture/#architecture-of-an-exchange), which monitors for deposits and updates the Balances database. 
Run the code below to have the Exchange server listen for deposit transactions and confirm the deposits.

> On a blockchain, multiple [validators](h[ttps://docs.flare.network/tech/glossary/](https://docs.flare.network/tech/validators/)) examine newly created transactions and come to an agreement about whether they are valid. 

Reviewer: ```suggestion
On a blockchain, multiple [validators](https://docs.flare.network/tech/validators/) examine newly created transactions and come to an agreement about whether they are valid. 
```

Be careful!

Have you tried actually rendering the markdown to produce an HTML page? You would very easily see that this does not show up correctly.

Author: I will try that today....and add a correct link.

When they are valid, they are added to a new block on the chain and identified by a one-way cryptographic algorithm, called a _hash_. 

> The one-way hash prevents transactions from being reverted. 

Reviewer: Not quite... the hash is a _signature_, so what it does is prevent the block from being tampered with after it has been added to the blockchain.

Author: Got it. I wrote _hash_ out of the intro and focused on validators.

> [Blocks](https://docs.flare.network/tech/glossary/) are added to the [blockchain](https://docs.flare.network/tech/glossary/) in chronological order and it may take a few blocks for the hash to be completed.

Reviewer: Again, not exactly. Calculating a hash is immediate, what takes time is for all the validators to agree on which should be the next block to be added.
Think of a room full of people trying to agree on anything... it takes some time.

Author: Got it.

> To prevent problems, we confirm deposits several blocks back (five blocks in our code), when the blocks are most likely to be irreversible.

Reviewer: And the final piece of the puzzle. Waiting for 100% agreement of all validators would be too slow (imagine you pay with your credit card on a shop and you are waiting for the payment to be confirmed. You want this to be quick!).
Therefore, transactions are usually accepted **before** there's full consensus. This means that people on different parts of the room might reach different conclusions on whether your transaction was accepted or not. These are called different _branches_ of the blockchain.
Later, since these people are constantly moving around the room and talking to each other, some might realize that different conclusions were reached for a past block and they'll fix it. Fixing it means that one of the branches is discarded and its followers jump to the other one. And in turn, this means that some transactions that had already been accepted in the discarded branch are now not accepted anymore. THIS is where reverted transactions come from.
And the way to avoid this is to wait for a few blocks, so that all validators have had a chance to look at your transaction and you are sure it will not be reverted.

Author: Awesome, I'm learning some more detail about how validators work.

> Newly submitted deposits to our receiving address are labeled "pending."

Reviewer: I always try to avoid talking about "we", "us" or "our".

Author: I agree. Sometimes that takes us to passive voice, which is also evil.

After five blocks have been created, valid deposits are labeled "confirmed."

Run the javascript below to see pending and confirmed transactions and which block number they were added to.

## One-Time Setup

1. Install [node.js](https://nodejs.org/en/download/). 
On the command line, run `node -v`  to check your installation.

2. Clone the Flare repository: 

`git clone https://github.com/flare-foundation/flare.git

> cd flare`

Reviewer: Code snippets should use code blocks like this:

```sh
git clone https://github.com/flare-foundation/flare.git
cd flare
```

Single backquotes should be used only for inline literal text, when referring to `variables` or `program_names`, for example.

Author: Makes sense.

See [the Flare repository on GitHub](https://github.com/flare-foundation/flare).

3. Install the web3.js library: 

`npm install web3`

See [Adding web3.js](https://web3js.readthedocs.io).

4. Build Flare with this script:

`./scripts/build.sh`

## Run the Code

1. `cd` to `flare`

2. Run the code: 

The following code:

* Sets up the deposit by instantiating `web3`, assigning the endpoint URL, and assigning the `receivingAddress`

> * Subscribes to 'pending transactions` to listen for deposit transactions and, if they are sent to the receiving address, labels them as "pending."

Reviewer: Be careful, the quotes used around `pending transactions` do not match so they break the format. Again, rendering the markdown into HTML would easily show you that there is a problem.

Author: Got it.

    let tx = await web3.eth.getTransaction(transactionHash);
    // If it is directed to our address...
    if (tx.to === receivingAddress) {
        // mark it as pending.
        console.log("Transaction", tx.hash, "is pending");
    }
}).on("error", console.error);

// Retrieve blocks 5 blocks back for confirmation.
web3.eth.subscribe("newBlockHeaders")
// When a new block has been produced...
.on("data", async (blockHeader) => {
    // retrieve a block old enough to be considered confirmed.
    let block = await web3.eth.getBlock(blockHeader.number - 5);

    // Get all of its transactions.
    block.transactions.forEach(async (transactionHash) => {
        // Retrieve the transaction...
        let tx = await web3.eth.getTransaction(transactionHash);
        // If it is directed to our address...
        if (tx.to === receivingAddress) {
            // mark it as confirmed.
            console.log("Transaction", tx.hash,
                "is confirmed in block", block.number);
        }
    });
}).on("error", console.error);
```

### Outcome

You can see which transactions are pending and which are confirmed and the block number they are in.

For example,

> ```javascript

Reviewer: Well, this code block is not javascript. But we're not strict about this, this is only used to produce syntax coloring so you are free to use whatever "looks good".

Author: Obvioulsy not a language I know. What is it? Typescript?

Reviewer: It's isn't any language, just console output. So the language tag should be just `text`, meaning that no syntax highlighting will be used.
That said, sometimes, if the console output _resembles_ some other language (like js), it's worth using `javascript` instead of `text` so you get the numbers colored differently, for example, which is nice.

Author: Got it. So I:
* changed mention of the code to "code," instead of "javascript."
* Gave the file suggestion as <name-of-file>.txt (It runs fine.)
* But I think "javascript" for the code block and outcome helps the reader (especially the grayed-out comments).

Reviewer: Oh, we didn't understand each other! The code starting in line 61 is definitely javascript!
What is not javascript is the console output starting in line 120, that's what my initial comment referred to.
So you can rename the file to `.js`. `.txt` will work, but it's misleading.

Reviewer: 👍

# Confirm Deposits to the Exchange

Deposits are made in several stages. 
After the user deposits funds to their assigned reception wallet (step 1 in [Detecting Deposits](https://docs.flare.network/exchange/architecture/#detecting-deposits)), the [Exchange server](https://docs.flare.network/exchange/architecture/#architecture-of-an-exchange) monitors for deposits and updates the Balances database. 
Run the code below to have the Exchange server listen for deposit transactions and confirm the deposits.

On a [blockchain](https://docs.flare.network/tech/glossary/), multiple [validators](https://docs.flare.network/tech/validators/) examine newly created transactions and decide on whether they are valid and which [block](https://docs.flare.network/tech/glossary/) to add to the chain next. 

> It takes time for all the validators to agree. 

Reviewer: I know that's personal, but do you like short, disconnected sentences like this one? I prefer to connect them, e.g.: "Since it takes time for all validators to agree, there is a period of time in which they might add or revert transactions to a block."
This is up for discussion, obviously, but it needs to be agreed upon!

Author: Much better. (The short sentences are me figuring out the logic and what to leave out to keep it simple.)

Until they reach agreement, validators may add or revert transactions to a block. 

> After the validation process is complete, reversing a transaction is very unlikely. 

Author: I see from the page on validators that probabilities of being a leader are involved in the consensus algorithm. Is there a point at which transactions on a block become irreversible/immutable?

Reviewer: The description of the consensus you are giving here is generic, and I think that is OK. I don't think there's any need to go into the details of how the consensus works for this page. So no need to talk about leaders.

Regarding immutability of transactions, some blockchains implement "Finality", which is a process that guarantees that a transaction won't change. But not us (nor Ethereum, for instance), so there is no such guarantee.
ANY transaction can be reverted in the future, but the probability of that happening is reduced with each new block. Therefore after a certain number of blocks it is commonly accepted that the transaction is "secure". The exact number of blocks depends on the blockchain and the security you want to provide: it's a tradeoff between security and speed.
The example uses 5 blocks, I think Bitcoin and Ethereum use 6 blocks, and some Exchanges might use 12 blocks, just to be extra secure.

Author: Thanks.

Just to make sure (for my own understanding....no change in text), it sounds like you're saying that validators can revert a transaction, but it is understood that blocks are still tamper-resistant from malicious actors (by virtue of cryptography).

> To avoid the risk of confirming a block that could be reverted, confirm deposits several blocks back (for example, five blocks back as in the code below). 

Reviewer: ```suggestion
To avoid the risk of accepting a deposit from a block that could be later reverted, only transactions that are a few blocks old are considered (for example, five blocks back as in the code below). 
```

I think this is clearer? This code is not confirming any block, since this is not code to be run by a validator, this is code for an Exchange. We're just acting upon blocks here.

Author: Thanks for the nuance.

Older blocks are less likely to be reverted.

Newly submitted deposits to the receiving address are labeled "pending."

> After five more blocks have been created, valid deposits are labeled "confirmed."

Reviewer: ```suggestion
After five more blocks have been created, valid deposits are then labeled "confirmed."
```

Again, I like creating a story rather than having disconnected sentences. Up for debate.

Author: I was thinking that "after" clarifies the timeline, but developers' first language may be have them thinking it the "if-then" pattern. So, agreed. 
(In general, I try to avoid redundancy, but sometimes it is good for emphasis and helps the reader. Notice I use the phrase "your own" on line 65...that's to help them hear it more clearly.)

Run the code below to see pending and confirmed transactions and which block number they were added to.

## One-Time Setup

1. Install [node.js](https://nodejs.org/en/download/). 

   On the command line, run `node -v`  to check your installation.

2. Clone the Flare repository: 

   ```
   git clone https://github.com/flare-foundation/flare.git
   cd flare
   ```

   See [the Flare repository on GitHub](https://github.com/flare-foundation/flare).

3. Install the web3.js library: 

>    `npm install web3`

Reviewer: Remember to use single backquotes only in inline text. This is a standalone code block so it should use triple backquotes (as in the `git clone` block above).

Author: That makes sense! (My reviewers in the past have let me get away with this!) Much better.... Any standalone code, not just more than 1 line.

   Run `npm web3 -v`  to check your installation.

   See [Adding web3.js](https://web3js.readthedocs.io).

4. Build Flare with this script:

   `./scripts/build.sh`

## Run the Code

The following code:

> * Sets up the deposit by instantiating `web3`, assigning the endpoint URL, and assigning the `receivingAddress`

Reviewer: Missing period.

Author: I went back and forth about that. Sometimes I leave a period out after code, so it doesn't get confused with the code. In this case it's really clear, so agreed.

Author: I went back and forth about that. Sometimes I leave a period out after code, so it doesn't get confused with the code, but it's really clear here, so agreed.

> * Subscribes to `pending transactions` to listen for deposit transactions and, if they are sent to the receiving address, labels them as "pending."

Reviewer: It's actually `pendingTransactions`. The backquotes mean this is literal text, so it should match the code exactly.

Author: Typo sorry.
